src/databricks/labs/ucx/mixins/cached_workspace_path.py

```diff
@@ -1,139 +1,138 @@
-                msg = f"Cannot initialise without object path: {object_info}"
-                raise ValueError(msg)
-            child = _CachedPath(self._cache, self._ws, path)
-            yield child.with_object_info(object_info)
-
-    def open(
+            yield self._from_object_info_with_cache(self._cache, self._ws, object_info)
+
+    @classmethod
+    def _detect_encoding_bom(cls, data: io.BytesIO) -> str | None:
+        # Peek at the first (up to) 4 bytes.
+        maybe_bom = data.read(4)
+        data.seek(-len(maybe_bom), os.SEEK_CUR)
+        if maybe_bom.startswith(codecs.BOM_UTF32_LE) or maybe_bom.startswith(codecs.BOM_UTF32_BE):
+            return "utf-32"
+        if maybe_bom.startswith(codecs.BOM_UTF16_LE) or maybe_bom.startswith(codecs.BOM_UTF16_BE):
+            return "utf-16"
+        if maybe_bom.startswith(codecs.BOM_UTF8):
+            return "utf-8"
+        return None
+
+    def open(  # type: ignore[override]
         self,
         mode: str = "r",
         buffering: int = -1,
         encoding: str | None = None,
         errors: str | None = None,
         newline: str | None = None,
-    ):
-        # only cache reads
-        if 'r' in mode:
-            return self._cache.open(self, mode, buffering, encoding, errors, newline)
-        self._cache.remove(str(self))
-        return super().open(mode, buffering, encoding, errors, newline)
+    ) -> IO:
+        # We only cache reads; if a write happens we use the default implementation (and evict any cache entry).
+        if 'w' in mode:
+            self._cache.remove(self)
+            return super().open(mode, buffering, encoding, errors, newline)
+
+        binary_data = self._cache.load(self)
+        binary_io = BytesIO(binary_data)
+        if 'b' in mode:
+            return binary_io
 
-    def _cached_open(self, mode: str, buffering: int, encoding: str | None, errors: str | None, newline: str | None):
-        return super().open(mode, buffering, encoding, errors, newline)
+        use_encoding = self._detect_encoding_bom(binary_io) if encoding is None else encoding
+        return io.TextIOWrapper(binary_io, use_encoding, errors, newline)
 
     # _rename calls unlink so no need to override it
     def unlink(self, missing_ok: bool = False) -> None:
-        self._cache.remove(str(self))
+        self._cache.remove(self)
         return super().unlink(missing_ok)
 
 
+_CachedPathT = TypeVar("_CachedPathT", bound=_CachedPath)
+
+
 class WorkspaceCache:
 
-    def __init__(self, ws: WorkspaceClient, max_entries=2048):
+    def __init__(self, ws: WorkspaceClient, max_entries: int = 2048) -> None:
         self._ws = ws
         self._cache = _PathLruCache(max_entries)
 
-    def get_path(self, path: str):
+    def get_workspace_path(self, path: str) -> WorkspacePath:
+        if not path.startswith("/"):
+            msg = f"Invalid workspace path; must be absolute and start with a slash ('/'): {path}"
+            raise ValueError(msg)
```

Comment:
please catch this exception in all places in production code that call this method and return `MaybeGraph(None, [DependencyProblem(...)])`

Reply:
I've updated all the call paths that (indirectly) call this method, and convert the exception raised here into `DependencyProblem`.

To avoid accidentally catching other triggers for `ValueError`, I updated it to a specific subclass.